fix(sandbox): Output to JSON

ddd-mtl opened the pull request:
### Summary

Fix for [#5006](https://github.com/holochain/holochain/issues/5006)

Changed the output of `hc s call [...]` to always output valid JSON when relevant data is returned.
Output format is not changed for calls returning a simple success/error confirmation message.
The JSON output is not "pretty printed" as we assume that this will be done by other tools like `jq`.

### Details

Standardized all JSON output to be made by calling `println!("{}", serde_json::to_value(data)?);`

When calling `ListApps` and  `InstallApp`, the JSON of `AppInfo` is post-processed to convert `agent_pub_key` and `cell_id` fields to base64. Other data fields are not changed.

The output of `ListDnas`, `ListCells`, has been converted to base64.

For outputting the result of `ListAgents`, this PR implements an `AgentResponse` struct that is serialized to JSON.
The way the fields are computed has not been changed.

### Example Output: InstallApp

#### Current output: Debug
```
AppInfo { installed_app_id: "my-fixture-app-deferred", cell_info: {"role-1": [Provisioned(ProvisionedCell { cell_id: CellId(DnaHash(uhC0knZTJ6H94hSYstUDXm-10EZTc_3XuharumZg_Zj7c2VezLOnE), AgentPubKey(uhCAkPlw2rA2F6TWxzRgelgkRE_nd8Why77W_RNMokmdrX4rVWRcE)), dna_modifiers: DnaModifiers { network_seed: "0123456", properties: null }, name: "a dna" })]}, status: AwaitingMemproofs, agent_pub_key: AgentPubKey(uhCAkPlw2rA2F6TWxzRgelgkRE_nd8Why77W_RNMokmdrX4rVWRcE), manifest: V0(AppManifestV0 { name: "my-fixture-app-deferred", description: None, roles: [AppRoleManifest { name: "role-1", provisioning: Some(Create { deferred: false }), dna: AppRoleDnaManifest { path: Some("a dna.dna"), modifiers: DnaModifiersOpt { network_seed: Some("0123456"), properties: None }, installed_hash: Some(HoloHashB64(DnaHash(uhC0knZTJ6H94hSYstUDXm-10EZTc_3XuharumZg_Zj7c2VezLOnE))), clone_limit: 0 } }], allow_deferred_memproofs: true }), installed_at: Timestamp(2025-07-21T14:52:31.617618Z) }
```


#### As plain JSON

```
{"installed_app_id":"my-fixture-app-deferred","cell_info":{"role-1":[{"type":"provisioned","value":{"cell_id":[[132,45,36,157,148,201,232,127,120,133,38,44,181,64,215,155,237,116,17,148,220,255,117,238,133,170,238,153,152,63,102,62,220,217,87,179,44,233,196],[132,32,36,62,92,54,172,13,133,233,53,177,205,24,30,150,9,17,19,249,221,241,104,114,239,181,191,68,211,40,146,103,107,95,138,213,89,23,4]],"dna_modifiers":{"network_seed":"0123456","properties":[192]},"name":"a dna"}}]},"status":{"type":"awaiting_memproofs"},"agent_pub_key":[132,32,36,62,92,54,172,13,133,233,53,177,205,24,30,150,9,17,19,249,221,241,104,114,239,181,191,68,211,40,146,103,107,95,138,213,89,23,4],"manifest":{"manifest_version":"0","name":"my-fixture-app-deferred","description":null,"roles":[{"name":"role-1","provisioning":{"strategy":"create","deferred":false},"dna":{"path":"a dna.dna","modifiers":{"network_seed":"0123456","properties":null},"installed_hash":"uhC0knZTJ6H94hSYstUDXm-10EZTc_3XuharumZg_Zj7c2VezLOnE","clone_limit":0}}],"allow_deferred_memproofs":true},"installed_at":1753109551617618}
```



#### Proposed output: post-processed JSON


```
{"installed_app_id":"my-fixture-app-deferred","cell_info":{"role-1":[{"type":"provisioned","value":{"cell_id":{"dna_hash":"uhC0knZTJ6H94hSYstUDXm-10EZTc_3XuharumZg_Zj7c2VezLOnE","agent_pub_key":"uhCAkeoB2vx_IF1tZguEeUv-0DLs2eonBQF8z5vyOBHjCm54hu2Gq"},"dna_modifiers":{"network_seed":"0123456","properties":[192]},"name":"a dna"}}]},"status":{"type":"awaiting_memproofs"},"agent_pub_key":"uhCAkeoB2vx_IF1tZguEeUv-0DLs2eonBQF8z5vyOBHjCm54hu2Gq","manifest":{"manifest_version":"0","name":"my-fixture-app-deferred","description":null,"roles":[{"name":"role-1","provisioning":{"strategy":"create","deferred":false},"dna":{"path":"a dna.dna","modifiers":{"network_seed":"0123456","properties":null},"installed_hash":"uhC0knZTJ6H94hSYstUDXm-10EZTc_3XuharumZg_Zj7c2VezLOnE","clone_limit":0}}],"allow_deferred_memproofs":true},"installed_at":1753201498337207}

```

### TODO:
- [ ] CHANGELOGs updated with appropriate info
- [x] All code changes are reflected in docs, including module-level docs

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * CLI commands now output structured JSON with base64-encoded identifiers for improved interoperability.
  * Agent listing and metadata commands provide JSON-formatted responses with enhanced detail.

* **Bug Fixes**
  * Improved extraction of agent public keys in test outputs by parsing JSON directly.

* **Documentation**
  * Updated README for clearer explanations and accurate command descriptions.

* **Style**
  * JSON outputs are now compact, removing unnecessary whitespace for easier parsing.

* **Tests**
  * Added and updated tests to validate new JSON serialization logic and output formats.

* **Chores**
  * Updated dependencies and development tools to support new features and testing.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ddd-mtl commented on crates/hc_sandbox/src/calls.rs:
```diff
@@ -393,64 +393,78 @@ async fn call_inner(client: &mut AdminWebsocket, call: AdminRequestCli) -> anyho
                 .uninstall_app(args.app_id.clone(), args.force)
                 .await?;
-            msg!("Uninstalled app: {}", args.app_id,);
+            msg!("Uninstalled app: \"{}\"", args.app_id);
```
Added quotes to the displayed app_id

ddd-mtl commented on crates/hc_sandbox/src/calls.rs:
```diff
@@ -393,64 +393,78 @@ async fn call_inner(client: &mut AdminWebsocket, call: AdminRequestCli) -> anyho
         AdminRequestCli::EnableApp(args) => {
             client.enable_app(args.app_id.clone()).await?;
-            msg!("Activated app: {:?}", args.app_id);
+            msg!("Enabled app: \"{}\"", args.app_id);
```
Changed wording from `Activated` to `Enabled`, as to be consistent with API names (e.g. `Enable App`)

ddd-mtl commented on crates/hc_sandbox/src/calls.rs:
```diff
@@ -528,35 +535,135 @@ async fn call_inner(client: &mut AdminWebsocket, call: AdminRequestCli) -> anyho
             let string_key_info = info
                 .into_iter()
-                .map(|(k, v)| (DnaHashB64::from(k).to_string(), v))
+                .map(|(k, v)| (k.to_string(), v))
```
`DnaHash` automatically converts to base64 when calling `to_string()`

jost-s commented on crates/hc_sandbox/src/calls.rs:
```diff
@@ -570,35 +580,135 @@ async fn call_inner(client: &mut AdminWebsocket, call: AdminRequestCli) -> anyho
 }
 
+/// Convert an [AppInfo] to JSON with extra base64 conversion of `agent_pub_key` and `cell_id` fields.
+fn convert_app_info(app_info: AppInfo) -> anyhow::Result<serde_json::Value> {
```
I suggest to include the target format in the function names
```suggestion
fn app_info_to_json(app_info: AppInfo) -> anyhow::Result<serde_json::Value> {
```

jost-s commented on crates/hc_sandbox/src/calls.rs:
```diff
@@ -570,35 +580,135 @@ async fn call_inner(client: &mut AdminWebsocket, call: AdminRequestCli) -> anyho
+}
+
+/// Convert the inner `cell_id` fields of the JSON value of a `IndexMap<RoleName, Vec<CellInfo>>`.
+fn transform_cell_info_map(value: serde_json::Value) -> anyhow::Result<serde_json::Value> {
```
```suggestion
fn cell_info_map_to_json(value: serde_json::Value) -> anyhow::Result<serde_json::Value> {
```

ddd-mtl replied:
Here the input is already JSON, so an explicit function name would rather be `convert_cell_id_to_base64_within_cell_info_map()`

jost-s commented on crates/hc_sandbox/src/calls.rs:
```diff
@@ -570,35 +580,135 @@ async fn call_inner(client: &mut AdminWebsocket, call: AdminRequestCli) -> anyho
+    Ok(serde_json::Value::Object(cell_info_map_map))
+}
+
+fn transform_cell_id(value: serde_json::Value) -> anyhow::Result<serde_json::Value> {
```
etc.

jost-s reviewed (approved):
A suggestion for the conversion function names. It would be great if you posted an example of the output (at least the new one, or even the old and the new ones in comparison) in the PR.

Looks good to me

ddd-mtl commented:
@jost-s Do these names work for you? https://github.com/holochain/holochain/pull/5142/commits/97c6e10bdd5df964ffe6758213d7b1f87e8125a4

I've also added an output example in the PR description.

ThetaSinner commented on crates/hc_sandbox/src/calls.rs:
```diff
@@ -418,64 +417,82 @@ async fn call_inner(client: &mut AdminWebsocket, call: AdminRequestCli) -> anyho
-            msg!("Added agent {}", agent);
+            println!("{}", serde_json::to_value(agent.to_string())?);
         }
         AdminRequestCli::ListCells => {
```
The display of this isn't clear, I'm seeing

```shell
[
  [
    "uhC0kQdsfrTwLmRR26atXw-cym6Euz0D178ETfKyHjRQ1hT6mNosd",
    "uhCAkErn6N1kBYVHHhbPq_KySEOhhW_s_Ffd91ryN1fxcdYVqfo8E"
  ]
]
```

Previously this would have looked like `DnaHash("uhC0kQdsfrTwLmRR26atXw-cym6Euz0D178ETfKyHjRQ1hT6mNosd")` and similar for the agent key. I know the JS client translates the type this way but if it's going to be transformed for display on the command line then I think making it a structured type with field names would be easier to read.

ddd-mtl replied:
yeah it depends how much the user already knows how a CellId is structured. I can change and have explicit fields instead.

ddd-mtl replied:
I changed the output example

ThetaSinner commented on crates/hc_sandbox/src/calls.rs:
```diff
@@ -418,64 +417,82 @@ async fn call_inner(client: &mut AdminWebsocket, call: AdminRequestCli) -> anyho
+            let converted = cell_id_values
+                .into_iter()
+                .map(cell_id_json_to_base64_json)
+                .collect::<anyhow::Result<Vec<serde_json::Value>>>()?;
```
This is library code. We used to use `anyhow` pretty heavily in library code but it is needlessly hard to integrate with other error handling libraries when this code is used by other applications. Prefer using an existing error type where possible, or use `thiserror` to create a new error type otherwise.

ThetaSinner commented on crates/hc_sandbox/src/calls.rs:
```diff
@@ -418,64 +417,82 @@ async fn call_inner(client: &mut AdminWebsocket, call: AdminRequestCli) -> anyho
         AdminRequestCli::DumpConductorState => {
             let state = client.dump_conductor_state().await?;
-            msg!("DUMP CONDUCTOR STATE \n{}", state);
+            println!("{}", state);
```
This is coming out pretty printed, is that happening in the conductor? That looks like a bug now that we're making the output consistently unformatted JSON.

ddd-mtl replied:
I can change that in `Conductor::dump_conductor_state()`

ThetaSinner commented on crates/hc_sandbox/src/calls.rs:
```diff
@@ -484,8 +501,8 @@ async fn call_inner(client: &mut AdminWebsocket, call: AdminRequestCli) -> anyho
                 "{}",
-                serde_json::to_string(
-                    &metrics
+                serde_json::to_value(
```
I now see this is consistent with above but why convert to value and then to string?

ddd-mtl replied:
`serde_json::to_string()` internally converts to value first so since we need to use  `println!()` anyway which is going to convert to string , the only necessary part we need is `to_value()`. This makes a separation of concerns more explicit: if `to_value()` fails we know its exclusively related to the conversion to JSON and not to string.

ThetaSinner replied:
As detail for this, the method that's invoked to do the serializatin inside the library is this one https://docs.rs/serde_json/1.0.141/serde_json/ser/fn.to_writer.html

That explicitly documents that the input structure is serialized to the output stream.

ThetaSinner replied:
I'd prefer this wasn't a case of belief. The absolute most efficient way to do this is with `to_writer` and providing a handle to stdout but I think `to_string` is a reasonable middle ground of not making the code too clever and still being reasonably efficient. I think it's important that we can come to an agreement about things like that.

ThetaSinner commented on crates/hc_sandbox/src/calls.rs:
```diff
@@ -570,35 +580,139 @@ async fn call_inner(client: &mut AdminWebsocket, call: AdminRequestCli) -> anyho
 }
 
+/// Convert an [`AppInfo`] to JSON with extra base64 conversion of `agent_pub_key` and `cell_id` fields.
+fn app_info_to_json(app_info: AppInfo) -> anyhow::Result<serde_json::Value> {
```
JSON conversions against types that can change need tests please

ThetaSinner reviewed (changes requested):
Requesting unit tests for the hash conversion logic and a few other queries

ddd-mtl commented on crates/hc_sandbox/README.md:
````diff
@@ -162,23 +163,26 @@ hc sandbox call --help
 ```text
 [... options and flags ...]
````
Agree. I'll remove it.

ddd-mtl commented on crates/hc_sandbox/src/calls.rs:
```diff
@@ -418,64 +417,82 @@ async fn call_inner(client: &mut AdminWebsocket, call: AdminRequestCli) -> anyho
                 }])
                 .await?;
-            msg!("Added admin port {}", port);
+            println!("{}", serde_json::to_value(port)?);
```
Indeed. it was for consistency, and since this is structured data, I'll revert the change and keep the initial messsage.

ddd-mtl commented:
Added unit-tests to base64 json conversion functions.
Removed pretty printing from conductor state dump.
Implemented a `CellIdJson` conversion struct in order to display it as fields.
Removed redundant and out of date documentation.
